// @sentry/mcp-server/dist/create-dsn-Bgwi-Ygi.cjs
const require_chunk = require('./chunk-DWy1uDak.cjs');
const require_defineTool = require('./defineTool-lWNTzWl_.cjs');
const require_api_utils = require('./api-utils-DFWUfQ83.cjs');
const require_schema = require('./schema-BL3kwfOA.cjs');
const zod = require_chunk.__toESM(require("zod"));
const __sentry_core = require_chunk.__toESM(require("@sentry/core"));

//#region src/tools/create-dsn.ts
var create_dsn_default = require_defineTool.defineTool({
	name: "create_dsn",
	description: [
		"Create an additional DSN for an EXISTING project.",
		"",
		"🔍 USE THIS TOOL WHEN:",
		"- Project already exists and needs additional DSN",
		"- 'Create another DSN for project X'",
		"- 'I need a production DSN for existing project'",
		"",
		"❌ DO NOT USE for new projects (use create_project instead)",
		"",
		"Be careful when using this tool!",
		"",
		"<examples>",
		"### Create additional DSN for existing project",
		"```",
		"create_dsn(organizationSlug='my-organization', projectSlug='my-project', name='Production')",
		"```",
		"</examples>",
		"",
		"<hints>",
		"- If the user passes a parameter in the form of name/otherName, its likely in the format of <organizationSlug>/<projectSlug>.",
		"- If any parameter is ambiguous, you should clarify with the user what they meant.",
		"</hints>"
	].join("\n"),
	inputSchema: {
		organizationSlug: require_schema.ParamOrganizationSlug,
		regionUrl: require_schema.ParamRegionUrl.optional(),
		projectSlug: require_schema.ParamProjectSlug,
		name: zod.z.string().trim().describe("The name of the DSN to create, for example 'Production'.")
	},
	async handler(params, context) {
		const apiService = require_api_utils.apiServiceFromContext(context, { regionUrl: params.regionUrl });
		const organizationSlug = params.organizationSlug;
		(0, __sentry_core.setTag)("organization.slug", organizationSlug);
		(0, __sentry_core.setTag)("project.slug", params.projectSlug);
		const clientKey = await apiService.createClientKey({
			organizationSlug,
			projectSlug: params.projectSlug,
			name: params.name
		});
		let output = `# New DSN in **${organizationSlug}/${params.projectSlug}**\n\n`;
		output += `**DSN**: ${clientKey.dsn.public}\n`;
		output += `**Name**: ${clientKey.name}\n\n`;
		output += "# Using this information\n\n";
		output += "- The `SENTRY_DSN` value is a URL that you can use to initialize Sentry's SDKs.\n";
		return output;
	}
});

//#endregion
Object.defineProperty(exports, 'create_dsn_default', {
  enumerable: true,
  get: function () {
    return create_dsn_default;
  }
});
