// @sentry/mcp-server/dist/create-team-CSBXDcoD.cjs
const require_chunk = require('./chunk-DWy1uDak.cjs');
const require_defineTool = require('./defineTool-lWNTzWl_.cjs');
const require_api_utils = require('./api-utils-DFWUfQ83.cjs');
const require_schema = require('./schema-BL3kwfOA.cjs');
const zod = require_chunk.__toESM(require("zod"));
const __sentry_core = require_chunk.__toESM(require("@sentry/core"));

//#region src/tools/create-team.ts
var create_team_default = require_defineTool.defineTool({
	name: "create_team",
	description: [
		"Create a new team in Sentry.",
		"",
		"🔍 USE THIS TOOL WHEN USERS WANT TO:",
		"- 'Create a new team'",
		"- 'Set up a team called [X]'",
		"- 'I need a team for my project'",
		"",
		"Be careful when using this tool!",
		"",
		"<examples>",
		"### Create a new team",
		"```",
		"create_team(organizationSlug='my-organization', name='the-goats')",
		"```",
		"</examples>",
		"",
		"<hints>",
		"- If any parameter is ambiguous, you should clarify with the user what they meant.",
		"</hints>"
	].join("\n"),
	inputSchema: {
		organizationSlug: require_schema.ParamOrganizationSlug,
		regionUrl: require_schema.ParamRegionUrl.optional(),
		name: zod.z.string().trim().describe("The name of the team to create.")
	},
	async handler(params, context) {
		const apiService = require_api_utils.apiServiceFromContext(context, { regionUrl: params.regionUrl });
		const organizationSlug = params.organizationSlug;
		(0, __sentry_core.setTag)("organization.slug", organizationSlug);
		const team = await apiService.createTeam({
			organizationSlug,
			name: params.name
		});
		let output = `# New Team in **${organizationSlug}**\n\n`;
		output += `**ID**: ${team.id}\n`;
		output += `**Slug**: ${team.slug}\n`;
		output += `**Name**: ${team.name}\n`;
		output += "# Using this information\n\n";
		output += `- You should always inform the user of the Team Slug value.\n`;
		return output;
	}
});

//#endregion
Object.defineProperty(exports, 'create_team_default', {
  enumerable: true,
  get: function () {
    return create_team_default;
  }
});
