// @sentry/mcp-server/dist/find-teams-BylQQ4FO.js
import { UserInputError } from "./errors-D_R568zP.js";
import { defineTool } from "./defineTool-77H4nSna.js";
import { apiServiceFromContext } from "./api-utils-BlUSO8xh.js";
import { ParamOrganizationSlug, ParamRegionUrl } from "./schema-DMfkTR1l.js";
import { setTag } from "@sentry/core";

//#region src/tools/find-teams.ts
var find_teams_default = defineTool({
	name: "find_teams",
	description: [
		"Find teams in an organization in Sentry.",
		"",
		"Use this tool when you need to:",
		"- View all teams in a Sentry organization",
		"- Find a team's slug to aid other tool requests"
	].join("\n"),
	inputSchema: {
		organizationSlug: ParamOrganizationSlug,
		regionUrl: ParamRegionUrl.optional()
	},
	async handler(params, context) {
		const apiService = apiServiceFromContext(context, { regionUrl: params.regionUrl });
		const organizationSlug = params.organizationSlug;
		if (!organizationSlug) throw new UserInputError("Organization slug is required. Please provide an organizationSlug parameter.");
		setTag("organization.slug", organizationSlug);
		const teams = await apiService.listTeams(organizationSlug);
		let output = `# Teams in **${organizationSlug}**\n\n`;
		if (teams.length === 0) {
			output += "No teams found.\n";
			return output;
		}
		output += teams.map((team) => `- ${team.slug}\n`).join("");
		return output;
	}
});

//#endregion
export { find_teams_default };
