// @sentry/mcp-server/dist/find-teams-DXyzyQOy.cjs
const require_chunk = require('./chunk-DWy1uDak.cjs');
const require_errors = require('./errors-BtJEw2Z-.cjs');
const require_defineTool = require('./defineTool-lWNTzWl_.cjs');
const require_api_utils = require('./api-utils-DFWUfQ83.cjs');
const require_schema = require('./schema-BL3kwfOA.cjs');
const __sentry_core = require_chunk.__toESM(require("@sentry/core"));

//#region src/tools/find-teams.ts
var find_teams_default = require_defineTool.defineTool({
	name: "find_teams",
	description: [
		"Find teams in an organization in Sentry.",
		"",
		"Use this tool when you need to:",
		"- View all teams in a Sentry organization",
		"- Find a team's slug to aid other tool requests"
	].join("\n"),
	inputSchema: {
		organizationSlug: require_schema.ParamOrganizationSlug,
		regionUrl: require_schema.ParamRegionUrl.optional()
	},
	async handler(params, context) {
		const apiService = require_api_utils.apiServiceFromContext(context, { regionUrl: params.regionUrl });
		const organizationSlug = params.organizationSlug;
		if (!organizationSlug) throw new require_errors.UserInputError("Organization slug is required. Please provide an organizationSlug parameter.");
		(0, __sentry_core.setTag)("organization.slug", organizationSlug);
		const teams = await apiService.listTeams(organizationSlug);
		let output = `# Teams in **${organizationSlug}**\n\n`;
		if (teams.length === 0) {
			output += "No teams found.\n";
			return output;
		}
		output += teams.map((team) => `- ${team.slug}\n`).join("");
		return output;
	}
});

//#endregion
Object.defineProperty(exports, 'find_teams_default', {
  enumerable: true,
  get: function () {
    return find_teams_default;
  }
});
