// @sentry/mcp-server/dist/handler-DTN2HYXt.js
import { defineTool } from "./defineTool-77H4nSna.js";
import { apiServiceFromContext, withApiErrorHandling } from "./api-utils-BlUSO8xh.js";
import { ParamOrganizationSlug, ParamRegionUrl } from "./schema-DMfkTR1l.js";
import { translateQuery } from "./agent-DZRRChrO.js";
import { formatIssueResults } from "./formatters-B3ONbRhI.js";
import { z } from "zod";
import { setTag } from "@sentry/core";

//#region src/tools/search-issues/handler.ts
var handler_default = defineTool({
	name: "search_issues",
	description: [
		"Search for grouped issues/problems in Sentry - returns a LIST of issues, NOT counts or aggregations.",
		"",
		"Uses AI to translate natural language queries into Sentry issue search syntax.",
		"Returns grouped issues with metadata like title, status, and user count.",
		"",
		"🔍 USE THIS TOOL WHEN USERS WANT:",
		"- A LIST of issues: 'show me issues', 'what problems do we have'",
		"- Filtered issue lists: 'unresolved issues', 'critical bugs'",
		"- Issues by impact: 'errors affecting more than 100 users'",
		"- Issues by assignment: 'issues assigned to me'",
		"",
		"❌ DO NOT USE FOR COUNTS/AGGREGATIONS:",
		"- 'how many errors' → use search_events",
		"- 'count of issues' → use search_events",
		"- 'total number of errors today' → use search_events",
		"- 'sum/average/statistics' → use search_events",
		"",
		"❌ ALSO DO NOT USE FOR:",
		"- Individual error events with timestamps → use search_events",
		"- Details about a specific issue ID → use get_issue_details",
		"",
		"REMEMBER: This tool returns a LIST of issues, not counts or statistics!",
		"",
		"<examples>",
		"search_issues(organizationSlug='my-org', naturalLanguageQuery='critical bugs from last week')",
		"search_issues(organizationSlug='my-org', naturalLanguageQuery='unhandled errors affecting 100+ users')",
		"search_issues(organizationSlug='my-org', naturalLanguageQuery='issues assigned to me')",
		"</examples>",
		"",
		"<hints>",
		"- If the user passes a parameter in the form of name/otherName, it's likely in the format of <organizationSlug>/<projectSlugOrId>.",
		"- Parse org/project notation directly without calling find_organizations or find_projects.",
		"- The projectSlugOrId parameter accepts both project slugs (e.g., 'my-project') and numeric IDs (e.g., '123456').",
		"</hints>"
	].join("\n"),
	inputSchema: {
		organizationSlug: ParamOrganizationSlug,
		naturalLanguageQuery: z.string().trim().min(1).describe("Natural language description of issues to search for"),
		projectSlugOrId: z.string().optional().describe("The project's slug or numeric ID (optional)"),
		regionUrl: ParamRegionUrl.optional(),
		limit: z.number().min(1).max(100).optional().default(10).describe("Maximum number of issues to return"),
		includeExplanation: z.boolean().optional().default(false).describe("Include explanation of how the query was translated")
	},
	async handler(params, context) {
		const apiService = apiServiceFromContext(context, { regionUrl: params.regionUrl });
		setTag("organization.slug", params.organizationSlug);
		if (params.projectSlugOrId) if (/^\d+$/.test(params.projectSlugOrId)) setTag("project.id", params.projectSlugOrId);
		else setTag("project.slug", params.projectSlugOrId);
		let projectId;
		if (params.projectSlugOrId) if (/^\d+$/.test(params.projectSlugOrId)) projectId = params.projectSlugOrId;
		else {
			const project = await withApiErrorHandling(() => apiService.getProject({
				organizationSlug: params.organizationSlug,
				projectSlugOrId: params.projectSlugOrId
			}), {
				organizationSlug: params.organizationSlug,
				projectSlugOrId: params.projectSlugOrId
			});
			projectId = String(project.id);
		}
		const translatedQuery = await translateQuery({
			naturalLanguageQuery: params.naturalLanguageQuery,
			organizationSlug: params.organizationSlug,
			projectSlugOrId: params.projectSlugOrId,
			projectId
		}, apiService);
		const issues = await withApiErrorHandling(() => apiService.listIssues({
			organizationSlug: params.organizationSlug,
			projectSlug: params.projectSlugOrId,
			query: translatedQuery.query,
			sortBy: translatedQuery.sort || "date",
			limit: params.limit
		}), {
			organizationSlug: params.organizationSlug,
			projectSlug: params.projectSlugOrId,
			query: translatedQuery.query
		});
		let output = "";
		if (params.includeExplanation) {
			output += `# Search Results for "${params.naturalLanguageQuery}"\n\n`;
			output += `⚠️ **IMPORTANT**: Display these issues as highlighted cards with status indicators, assignee info, and clickable Issue IDs.\n\n`;
			output += `## Query Translation\n`;
			output += `Natural language: "${params.naturalLanguageQuery}"\n`;
			output += `Sentry query: \`${translatedQuery.query}\``;
			if (translatedQuery.sort) output += `\nSort: ${translatedQuery.sort}`;
			if (translatedQuery.explanation) output += `\n\n${translatedQuery.explanation}`;
			output += `\n\n`;
			output += formatIssueResults(issues, params.organizationSlug, params.projectSlugOrId, translatedQuery.query, params.regionUrl, params.naturalLanguageQuery, true);
		} else output = formatIssueResults(issues, params.organizationSlug, params.projectSlugOrId, translatedQuery.query, params.regionUrl, params.naturalLanguageQuery, false);
		return output;
	}
});

//#endregion
export { handler_default };
