// @sentry/mcp-server/dist/promptDefinitions-DF8nuJVq.js
import { ParamIssueShortId, ParamIssueUrl, ParamOrganizationSlug } from "./schema-DMfkTR1l.js";
import { z } from "zod";

//#region src/promptDefinitions.ts
/**
* All MCP prompt definitions for the Sentry server.
*
* Used by server.ts to register prompts with the MCP server and by prompts.ts
* for implementation handlers. Each definition includes name, description, and
* Zod parameter schema for multi-step workflows.
*/
const PROMPT_DEFINITIONS = [{
	name: "find_errors_in_file",
	description: ["Use this prompt when you need to find errors in Sentry for a given file."].join("\n"),
	paramsSchema: {
		organizationSlug: ParamOrganizationSlug,
		filename: z.string().describe("The filename to search for errors in.")
	}
}, {
	name: "fix_issue_with_seer",
	description: ["Use this prompt when you need to fix an issue with Seer.", "You can pass in either an `issueId` and `organizationSlug`, or an `issueUrl`."].join("\n"),
	paramsSchema: {
		organizationSlug: ParamOrganizationSlug.optional(),
		issueId: ParamIssueShortId.optional(),
		issueUrl: ParamIssueUrl.optional()
	}
}];

//#endregion
export { PROMPT_DEFINITIONS };
