// @sentry/mcp-server/dist/promptDefinitions-CWV8hsUX.cjs
const require_chunk = require('./chunk-DWy1uDak.cjs');
const require_schema = require('./schema-BL3kwfOA.cjs');
const zod = require_chunk.__toESM(require("zod"));

//#region src/promptDefinitions.ts
/**
* All MCP prompt definitions for the Sentry server.
*
* Used by server.ts to register prompts with the MCP server and by prompts.ts
* for implementation handlers. Each definition includes name, description, and
* Zod parameter schema for multi-step workflows.
*/
const PROMPT_DEFINITIONS = [{
	name: "find_errors_in_file",
	description: ["Use this prompt when you need to find errors in Sentry for a given file."].join("\n"),
	paramsSchema: {
		organizationSlug: require_schema.ParamOrganizationSlug,
		filename: zod.z.string().describe("The filename to search for errors in.")
	}
}, {
	name: "fix_issue_with_seer",
	description: ["Use this prompt when you need to fix an issue with Seer.", "You can pass in either an `issueId` and `organizationSlug`, or an `issueUrl`."].join("\n"),
	paramsSchema: {
		organizationSlug: require_schema.ParamOrganizationSlug.optional(),
		issueId: require_schema.ParamIssueShortId.optional(),
		issueUrl: require_schema.ParamIssueUrl.optional()
	}
}];

//#endregion
Object.defineProperty(exports, 'PROMPT_DEFINITIONS', {
  enumerable: true,
  get: function () {
    return PROMPT_DEFINITIONS;
  }
});
