// @sentry/mcp-server/dist/schema-BL3kwfOA.cjs
const require_chunk = require('./chunk-DWy1uDak.cjs');
const require_constants = require('./constants-wrEbre3j.cjs');
const zod = require_chunk.__toESM(require("zod"));

//#region src/schema.ts
const ParamOrganizationSlug = zod.z.string().toLowerCase().trim().describe("The organization's slug. You can find a existing list of organizations you have access to using the `find_organizations()` tool.");
const ParamTeamSlug = zod.z.string().toLowerCase().trim().describe("The team's slug. You can find a list of existing teams in an organization using the `find_teams()` tool.");
const ParamProjectSlug = zod.z.string().toLowerCase().trim().describe("The project's slug. You can find a list of existing projects in an organization using the `find_projects()` tool.");
const ParamProjectSlugOrAll = zod.z.string().toLowerCase().trim().describe("The project's slug. This will default to all projects you have access to. It is encouraged to specify this when possible.");
const ParamIssueShortId = zod.z.string().toUpperCase().trim().describe("The Issue ID. e.g. `PROJECT-1Z43`");
const ParamIssueUrl = zod.z.string().url().trim().describe("The URL of the issue. e.g. https://my-organization.sentry.io/issues/PROJECT-1Z43");
const ParamPlatform = zod.z.string().toLowerCase().trim().describe("The platform for the project. e.g., python, javascript, react, etc.");
const ParamTransaction = zod.z.string().trim().describe("The transaction name. Also known as the endpoint, or route name.");
const ParamQuery = zod.z.string().trim().describe(`The search query to apply. Use the \`help(subject="query_syntax")\` tool to get more information about the query syntax rather than guessing.`);
/**
* Region URL parameter for Sentry API requests.
*
* Handles region-specific URLs for Sentry's Cloud Service while gracefully
* supporting self-hosted Sentry installations that may return empty regionUrl values.
* This schema accepts both valid URLs and empty strings to ensure compatibility
* across different Sentry deployment types.
*/
const ParamRegionUrl = zod.z.string().trim().refine((value) => !value || zod.z.string().url().safeParse(value).success, { message: "Must be a valid URL or empty string (for self-hosted Sentry)" }).describe("The region URL for the organization you're querying, if known. For Sentry's Cloud Service (sentry.io), this is typically the region-specific URL like 'https://us.sentry.io'. For self-hosted Sentry installations, this parameter is usually not needed and should be omitted. You can find the correct regionUrl from the organization details using the `find_organizations()` tool.");
const ParamIssueStatus = zod.z.enum([
	"resolved",
	"resolvedInNextRelease",
	"unresolved",
	"ignored"
]).describe("The new status for the issue. Valid values are 'resolved', 'resolvedInNextRelease', 'unresolved', and 'ignored'.");
const ParamAssignedTo = zod.z.string().trim().describe("The assignee in format 'user:ID' or 'team:ID' where ID is numeric. Example: 'user:123456' or 'team:789'. Use the whoami tool to find your user ID.");
const ParamSentryGuide = zod.z.enum(require_constants.SENTRY_GUIDES).describe("Optional guide filter to limit search results to specific documentation sections. Use either a platform (e.g., 'javascript', 'python') or platform/guide combination (e.g., 'javascript/nextjs', 'python/django').");
const ParamEventId = zod.z.string().trim().describe("The ID of the event.");
const ParamAttachmentId = zod.z.string().trim().describe("The ID of the attachment to download.");

//#endregion
Object.defineProperty(exports, 'ParamAssignedTo', {
  enumerable: true,
  get: function () {
    return ParamAssignedTo;
  }
});
Object.defineProperty(exports, 'ParamAttachmentId', {
  enumerable: true,
  get: function () {
    return ParamAttachmentId;
  }
});
Object.defineProperty(exports, 'ParamEventId', {
  enumerable: true,
  get: function () {
    return ParamEventId;
  }
});
Object.defineProperty(exports, 'ParamIssueShortId', {
  enumerable: true,
  get: function () {
    return ParamIssueShortId;
  }
});
Object.defineProperty(exports, 'ParamIssueStatus', {
  enumerable: true,
  get: function () {
    return ParamIssueStatus;
  }
});
Object.defineProperty(exports, 'ParamIssueUrl', {
  enumerable: true,
  get: function () {
    return ParamIssueUrl;
  }
});
Object.defineProperty(exports, 'ParamOrganizationSlug', {
  enumerable: true,
  get: function () {
    return ParamOrganizationSlug;
  }
});
Object.defineProperty(exports, 'ParamPlatform', {
  enumerable: true,
  get: function () {
    return ParamPlatform;
  }
});
Object.defineProperty(exports, 'ParamProjectSlug', {
  enumerable: true,
  get: function () {
    return ParamProjectSlug;
  }
});
Object.defineProperty(exports, 'ParamProjectSlugOrAll', {
  enumerable: true,
  get: function () {
    return ParamProjectSlugOrAll;
  }
});
Object.defineProperty(exports, 'ParamQuery', {
  enumerable: true,
  get: function () {
    return ParamQuery;
  }
});
Object.defineProperty(exports, 'ParamRegionUrl', {
  enumerable: true,
  get: function () {
    return ParamRegionUrl;
  }
});
Object.defineProperty(exports, 'ParamSentryGuide', {
  enumerable: true,
  get: function () {
    return ParamSentryGuide;
  }
});
Object.defineProperty(exports, 'ParamTeamSlug', {
  enumerable: true,
  get: function () {
    return ParamTeamSlug;
  }
});
Object.defineProperty(exports, 'ParamTransaction', {
  enumerable: true,
  get: function () {
    return ParamTransaction;
  }
});
