// @sentry/mcp-server/dist/tools-B0wbxDty.js
import { whoami_default } from "./whoami-Bswhb72a.js";
import { find_organizations_default } from "./find-organizations-Cs1GvZtJ.js";
import { find_teams_default } from "./find-teams-BylQQ4FO.js";
import { find_projects_default } from "./find-projects-CZCMck_0.js";
import { find_releases_default } from "./find-releases-Do_CywRr.js";
import { get_issue_details_default } from "./get-issue-details-_uS_AZ_v.js";
import { get_event_attachment_default } from "./get-event-attachment-CVoMrETq.js";
import { update_issue_default } from "./update-issue-DoQYEQjf.js";
import { handler_default } from "./handler-B0bLvFFX.js";
import { create_team_default } from "./create-team-D0dUl7Zo.js";
import { create_project_default } from "./create-project-Ca1Sy1_F.js";
import { update_project_default } from "./update-project-C9CLxNfF.js";
import { create_dsn_default } from "./create-dsn-BT3t-MDA.js";
import { find_dsns_default } from "./find-dsns-DrNAfU_Q.js";
import { analyze_issue_with_seer_default } from "./analyze-issue-with-seer-CpwtN4UA.js";
import { search_docs_default } from "./search-docs-DIm3ykOg.js";
import { get_doc_default } from "./get-doc-DyqJe6TP.js";
import { handler_default as handler_default$1 } from "./handler-DTN2HYXt.js";

//#region src/tools/index.ts
var tools_default = {
	whoami: whoami_default,
	find_organizations: find_organizations_default,
	find_teams: find_teams_default,
	find_projects: find_projects_default,
	find_releases: find_releases_default,
	get_issue_details: get_issue_details_default,
	get_event_attachment: get_event_attachment_default,
	update_issue: update_issue_default,
	search_events: handler_default,
	create_team: create_team_default,
	create_project: create_project_default,
	update_project: update_project_default,
	create_dsn: create_dsn_default,
	find_dsns: find_dsns_default,
	analyze_issue_with_seer: analyze_issue_with_seer_default,
	search_docs: search_docs_default,
	get_doc: get_doc_default,
	search_issues: handler_default$1
};

//#endregion
export { tools_default };
