// @sentry/mcp-server/dist/tools-qfk-QAo0.cjs
const require_whoami = require('./whoami-BOJHfdih.cjs');
const require_find_organizations = require('./find-organizations-BDtHkcFV.cjs');
const require_find_teams = require('./find-teams-DXyzyQOy.cjs');
const require_find_projects = require('./find-projects-Dh-1AK-m.cjs');
const require_find_releases = require('./find-releases-BsAkl--F.cjs');
const require_get_issue_details = require('./get-issue-details-C96fhps3.cjs');
const require_get_event_attachment = require('./get-event-attachment-BzhDeM6z.cjs');
const require_update_issue = require('./update-issue-DqUsAuTd.cjs');
const require_handler = require('./handler-7T1iMsQU.cjs');
const require_create_team = require('./create-team-CSBXDcoD.cjs');
const require_create_project = require('./create-project-4R4aZMoc.cjs');
const require_update_project = require('./update-project-DQu5PnE3.cjs');
const require_create_dsn = require('./create-dsn-Bgwi-Ygi.cjs');
const require_find_dsns = require('./find-dsns-CkvahHwJ.cjs');
const require_analyze_issue_with_seer = require('./analyze-issue-with-seer-ByoDTsq-.cjs');
const require_search_docs = require('./search-docs-EbjJuL9Q.cjs');
const require_get_doc = require('./get-doc-BfMg0GIa.cjs');
const require_handler$1 = require('./handler-DEjS3yks.cjs');

//#region src/tools/index.ts
var tools_default = {
	whoami: require_whoami.whoami_default,
	find_organizations: require_find_organizations.find_organizations_default,
	find_teams: require_find_teams.find_teams_default,
	find_projects: require_find_projects.find_projects_default,
	find_releases: require_find_releases.find_releases_default,
	get_issue_details: require_get_issue_details.get_issue_details_default,
	get_event_attachment: require_get_event_attachment.get_event_attachment_default,
	update_issue: require_update_issue.update_issue_default,
	search_events: require_handler.handler_default,
	create_team: require_create_team.create_team_default,
	create_project: require_create_project.create_project_default,
	update_project: require_update_project.update_project_default,
	create_dsn: require_create_dsn.create_dsn_default,
	find_dsns: require_find_dsns.find_dsns_default,
	analyze_issue_with_seer: require_analyze_issue_with_seer.analyze_issue_with_seer_default,
	search_docs: require_search_docs.search_docs_default,
	get_doc: require_get_doc.get_doc_default,
	search_issues: require_handler$1.handler_default
};

//#endregion
Object.defineProperty(exports, 'tools_default', {
  enumerable: true,
  get: function () {
    return tools_default;
  }
});
